src/osd/OSD.cc

```diff
@@ -2627,7 +2627,92 @@ int OSD::init()
   return 0;
 monout:
   mgrc.shutdown();
+  service.start_shutdown();
+  op_shardedwq.drain();
+  // Shutdown PGs
+  {
+    RWLock::RLocker l(pg_map_lock);
+    for (ceph::unordered_map<spg_t, PG*>::iterator p = pg_map.begin();
+        p != pg_map.end();
+        ++p) {
+      dout(20) << " kicking pg " << p->first << dendl;
+      p->second->lock();
+      p->second->on_shutdown();
+      p->second->unlock();
+      p->second->osr->flush();
+    }
+  }
+  clear_pg_stat_queue();
+  // drain op queue again (in case PGs requeued something)
+  op_shardedwq.drain();
+  {
+    finished.clear(); // zap waiters (bleh, this is messy)
+  }
+  op_shardedwq.clear_pg_slots();
+  osd_lock.Unlock();
+
+  heartbeat_lock.Lock();
+  heartbeat_stop = true;
+  heartbeat_cond.Signal();
+  heartbeat_lock.Unlock();
+  heartbeat_thread.join();
+
+  peering_tp.drain();
+  peering_wq.clear();
+  peering_tp.stop();
+  dout(10) << "osd tp stopped" << dendl;
+
+  osd_op_tp.drain();
+  osd_op_tp.stop();
+  dout(10) << "op sharded tp stopped" << dendl;
+
+  command_tp.drain();
+  command_tp.stop();
+  dout(10) << "command tp stopped" << dendl;
+
+  disk_tp.drain();
+  disk_tp.stop();
+  dout(10) << "disk tp paused (new)" << dendl;
+
+  dout(10) << "stopping agent" << dendl;
+  service.agent_stop();
+
+  osd_lock.Lock();
+
+  reset_heartbeat_peers();
+
+  tick_timer.shutdown();
+
+  {
+    Mutex::Locker l(tick_timer_lock);
+    tick_timer_without_osd_lock.shutdown();
+  }
+  service.shutdown_reserver();
+  pg_map.clear();
+  cct->_conf->remove_observer(this);
+  store->umount();
+  delete store;
+  store = 0;
+
   monc->shutdown();
+  osd_lock.Unlock();
+
+  osdmap = OSDMapRef();
+  service.shutdown();
+  op_tracker.on_shutdown();
+
+  class_handler->shutdown();
+  client_messenger->shutdown();
+  cluster_messenger->shutdown();
+  hb_front_client_messenger->shutdown();
+  hb_back_client_messenger->shutdown();
+  objecter_messenger->shutdown();
+  hb_front_server_messenger->shutdown();
+  hb_back_server_messenger->shutdown();
+
+  peering_wq.clear();
+  osd_lock.Lock(); // make locker happy
+  return r;
```

Comment:
On second thought, this is a bit excessive.. and is going to immediately break the next time we change any part of the shutdown procedure.  Is it possible to factor this into a helper shared with shutdown(), or should we instead just exit(0) or assert out?